### PR TITLE
API 2 stat logging and query handling

### PR DESCRIPTION
**Pull request by Wirewraith**
## About the PR
Datums what do stat logging and query handling. I'll flesh this description out further down the line.

## TODO
- [x] Finalise new datums
- [x] Place initialization logic in world/New or whatever
- [ ] Replace every single API call
- [ ] Replace every single stat log call

### CODE REVIEW

**Comment by Zonespace27** [comment]
Out of curiosity, will this have any effects on the frontend of goonhub or the like?

**Comment by Wirewraith** [comment]
> 
> 
> Out of curiosity, will this have any effects on the frontend of goonhub, or the like?

By itself, no. But API 2 is a large sprawling project that has other aspects that will result in a new goonhub site. Eventually.

File: code/datums/stat_logger/logger.dm [hunk 1]
```diff
@@ -0,0 +1,147 @@
+/* ._.-'~'-._.-'~'-._.-'~'-._.-'~'-._.-'~'-._. */
+/*-=-=-=-=-=-=-=STAT LOGGER BY WIRE-=-=-=-=-=-=*/
+/* '~'-._.-'~'-._.-'~'-._.-'~'-._.-'~'-._.-'~' */
+
+/// Holder datum for [/datum/stats]
+/var/global/datum/stats/stats
```
Comment (ZeWaka) [inline code/datums/stat_logger/logger.dm:6]:
```suggestion
var/global/datum/stats/stats
```
you also never initialize this

File: code/datums/stat_logger/logger.dm [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	/// A queue of stat records we couldn't save immediately (probably because we didn't have a round ID yet)
+	var/list/queue = list()
+	/// A list of foreign relationship definitions
+	var/list/foreign_relations = list()
+
+	New()
+		..()
+		// Build out all the foreign relationship definitions ahead of time
+		for (var/type in typesof(/datum/stats_foreign))
+			if (type == /datum/stats_foreign) continue
```
Comment (ZeWaka) [inline code/datums/stat_logger/logger.dm:35]: use `concrete_typesof(/datum/stats_foreign)`
and `ABSTRACT_TYPE(/datum/stats_foreign)`

File: code/datums/stat_logger/logger.dm (hunk 1 above)
Comment (Wirewraith) [reply code/datums/stat_logger/logger.dm:6]: One of the TODO items in the PR description is to place the initializing logic.

**Pull request closed without merging**